lte/gateway/c/session_manager/SessionStateEnforcer.h

```diff
@@ -61,52 +62,124 @@ class SessionStateEnforcer {
+      SessionMap& session_map, const std::string& imsi,
+      std::unique_ptr<SessionState>& session_state,
+      SessionUpdate& session_update);
   /*Charging & rule related*/
-  void handle_session_init_rule_updates(
+  bool handle_session_init_rule_updates(
       SessionMap& session_map, const std::string& imsi,
-      SessionState& session_state);
+      std::unique_ptr<SessionState>& session_state);
+
+  /* Move to idle state */
+  void m5g_move_to_inactive_state(
+      std::string& imsi, std::unique_ptr<SessionState>& session,
+      SetSmNotificationContext notif, SessionStateUpdateCriteria& uc);
+
+  /* Move to active state */
+  void m5g_move_to_active_state(
+      std::string& imsi, std::unique_ptr<SessionState>& session,
+      SetSmNotificationContext notif, SessionStateUpdateCriteria& uc);
 
   /*Release request handle*/
   bool m5g_release_session(
-      SessionMap& session_map, const std::string& imsi, const std::string& dnn,
+      SessionMap& session_map, const std::string& imsi, const uint32_t& pdu_id,
       SessionUpdate& session_update);
 
   /*Handle and update respective session upon receiving message from UPF*/
-  void m5g_update_session_state_to_amf(
+  void m5g_process_response_from_upf(
       const std::string& imsi, uint32_t teid, uint32_t version);
 
- private:
-  std::vector<std::string> static_rules;
+  /* Send session requst to upf */
+  void m5g_send_session_request_to_upf(
+      const std::string& imsi, const std::unique_ptr<SessionState>& session);
+
+  /*
+   * To send session or UE notifcaiton
+   * to AMF, this cud be session state or
+   * paging notification
+   */
+  void handle_state_update_to_amf(
+      SessionState& session_state, const magma::lte::M5GSMCause m5gsmcause,
+      NotifyUeEvents event);
+
+  /* Get next teid */
+  uint32_t get_next_teid();
+
+  /* Get current TEID */
+  uint32_t get_current_teid();
+
+  /* Get N3 ip  address of UPF */
+  std::string get_upf_n3_addr();
+
+  /* Get N3 ip  address of UPF */
+  std::string get_upf_node_id();
+
+  /* Initialize the upf node id and n3 address
+   */
+  bool set_upf_node(const std::string& node_id, const std::string& n3_addr);
 
+ private:
   ConvergedRuleStore GlobalRuleList;
   std::unordered_multimap<std::string, uint32_t> pdr_map_;
-  std::unordered_multimap<std::string, uint32_t> far_map_;
-
   std::shared_ptr<StaticRuleStore> rule_store_;
   SessionStore& session_store_;
   std::shared_ptr<PipelinedClient> pipelined_client_;
   std::shared_ptr<AmfServiceClient> amf_srv_client_;
   folly::EventBase* evb_;
   std::chrono::seconds retry_timeout_;
   magma::mconfig::SessionD mconfig_;
+  /* For now we are keeping local teid_counter in this class,
+   * in future once we support multiple UPF under SMF, below
+   * upf_node_id has to be moved/associated with PipelineClient,
+   * and the following 2 variables (counter + N3 ip address)
+   * has to be mainted for every PipeliendClient (UPF node_id)
+   */
+  std::string upf_node_id;

```

Comment:
nit: private variables should have trailing underscore

Reply:
Adopted changes.